File: zoo/src/main/scala/com/intel/analytics/zoo/models/recommendation/SessionRecommender.scala
/*
 * Copyright 2018 Analytics Zoo Authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.intel.analytics.zoo.models.recommendation

import com.intel.analytics.bigdl.dataset.{Sample, TensorSample}
import com.intel.analytics.bigdl.nn.Graph.ModuleNode
import com.intel.analytics.bigdl.nn.Sum
import com.intel.analytics.bigdl.nn.abstractnn.AbstractModule
import com.intel.analytics.bigdl.optim.LocalPredictor
import com.intel.analytics.bigdl.tensor.Tensor
import com.intel.analytics.bigdl.tensor.TensorNumericMath.TensorNumeric
import com.intel.analytics.bigdl.utils.Shape
import com.intel.analytics.zoo.models.common.ZooModel
import com.intel.analytics.zoo.pipeline.api.keras.layers._
import com.intel.analytics.zoo.pipeline.api.keras.models.Model
import org.apache.spark.rdd.RDD

import scala.reflect.ClassTag

/**
 * The session based recommender model.
 *
 * @param itemCount       The number of distinct items. Positive integer.
 * @param itemEmbed       The output size of embedding layer. Positive integer.
 * @param mlpHiddenLayers Units of hidden layers for the mlp model. Array of positive integers.
 * @param sessionLength   The max number of items in the sequence of a session
 * @param rnnHiddenLayers Units of hidden layers for the mlp model. Array of positive integers.
 * @param includeHistory  Whether to include purchase history. Boolean. Default is true.
 * @param historyLength   The max number of items in the sequence of historical purchase
 */
class SessionRecommender[T: ClassTag](
      val itemCount: Int,
      val itemEmbed: Int = 100,
      val rnnHiddenLayers: Array[Int] = Array(40, 20),
      val sessionLength: Int = 5,
      val includeHistory: Boolean = true,
      val mlpHiddenLayers: Array[Int] = Array(40, 20),
      val historyLength: Int = 10)(implicit ev: TensorNumeric[T]) extends Recommender[T] {

  override def buildModel(): AbstractModule[Tensor[T], Tensor[T], T] = {
    val inputRnn: ModuleNode[T] = Input(inputShape = Shape(sessionLength))

    // item embedding layer
    val sessionTable = Embedding[T](
      itemCount + 1, itemEmbed, "normal", inputLength = sessionLength).inputs(inputRnn)

    // rnn part
    var gru = GRU[T](rnnHiddenLayers(0), returnSequences = true).inputs(sessionTable)
    for (i <- 1 until rnnHiddenLayers.length - 1) {
      gru = GRU[T](rnnHiddenLayers(i), returnSequences = true).inputs(gru)
    }
    val gruLast = GRU[T](rnnHiddenLayers.last, returnSequences = false).inputs(gru)
    val rnn = Dense[T](itemCount).inputs(gruLast)

    // mlp part
    if (includeHistory) {
      val inputMlp: ModuleNode[T] = Input(inputShape = Shape(historyLength))
      val hisTable = Embedding[T](itemCount + 1, itemEmbed, inputLength = historyLength)
        .inputs(inputMlp)
      val sum = new KerasLayerWrapper[T](Sum[T](2)).inputs(hisTable)
      var mlp = Dense(mlpHiddenLayers(0), activation = "relu").inputs(sum)
      for (i <- 1 until mlpHiddenLayers.length) {
        mlp = Dense(mlpHiddenLayers(i), activation = "relu").inputs(mlp)
      }
      val mlpLast = Dense(itemCount).inputs(mlp)
      // combine rnn and mlp
      val merged = Merge.merge[T](List(mlpLast, rnn), "sum")
      val out = Activation("softmax").inputs(merged)
      Model(Array(inputMlp, inputRnn), out).asInstanceOf[AbstractModule[Tensor[T], Tensor[T], T]]
    }
    else {
      val out = Activation("softmax").inputs(rnn)
      Model(inputRnn, out).asInstanceOf[AbstractModule[Tensor[T], Tensor[T], T]]
    }
  }

  private def topk(_output: Tensor[T],
                   maxItems: Int,
                   zeroBasedLabel: Boolean = true): Array[(Int, Float)] = {

    val results: (Tensor[T], Tensor[T]) = _output.topk(maxItems, 1, false)

    val recommends = (1 to maxItems).map { i =>
      val zeroBase = if (zeroBasedLabel) 1 else 0
      val predict = (results._2.valueAt(i).asInstanceOf[Float] - zeroBase).toInt
      val probability = results._1.valueAt(i).asInstanceOf[Float]
      (predict, probability)
    }
    recommends.toArray
  }

  def recommendForSession(sessions: RDD[Sample[T]],
                          maxItems: Int,
                          zeroBasedLabel: Boolean): RDD[Array[(Int, Float)]] = {
    val raw = predict(sessions)
    raw.map(x => topk(x.toTensor[T], maxItems, zeroBasedLabel))
  }

  def recommendForSession(sessions: Array[Sample[T]],
                          maxItems: Int,
                          zeroBasedLabel: Boolean): Array[Array[(Int, Float)]] = {
    val raw = LocalPredictor(this).predict(sessions)
    raw.map(x => topk(x.toTensor[T], maxItems, zeroBasedLabel))
  }


  override def recommendForUser(featureRdd: RDD[UserItemFeature[T]],
                                maxItems: Int): RDD[UserItemPrediction] = {
    throw new UnsupportedOperationException(s"recommendForUser: Unsupported for SessionRecommender")
  }

  override def recommendForItem(featureRdd: RDD[UserItemFeature[T]],
                                maxItems: Int): RDD[UserItemPrediction] = {
    throw new UnsupportedOperationException(s"recommendForItem: Unsupported for SessionRecommender")
  }

  override def predictUserItemPair(featureRdd: RDD[UserItemFeature[T]]): RDD[UserItemPrediction] = {
    throw new UnsupportedOperationException(
      s"predictUserItemPair: Unsupported for SessionRecommender")
  }

}

object SessionRecommender {
  /**
   * The factory method to create a SessionRecommender instance.
   *
   * @param itemCount       The number of distinct items. Positive integer.
   * @param itemEmbed       The output size of embedding layer. Positive integer.
   * @param mlpHiddenLayers Units of hidden layers for the mlp model. Array of positive integers.
   * @param sessionLength       The max number of items in the sequence of a session
   * @param rnnHiddenLayers Units of hidden layers for the mlp model. Array of positive integers.
   * @param includeHistory  Whether to include purchase history. Boolean. Default is true.
   * @param historyLength       The max number of items in the sequence of historical purchase
   */
  def apply[@specialized(Float, Double) T: ClassTag](
      itemCount: Int,
      itemEmbed: Int = 100,
      rnnHiddenLayers: Array[Int] = Array(40, 20),
      sessionLength: Int = 5,
      includeHistory: Boolean = true,
      mlpHiddenLayers: Array[Int] = Array(40, 20),
      historyLength: Int = 10)(implicit ev: TensorNumeric[T]): SessionRecommender[T] = {
    new SessionRecommender[T](
      itemCount,
      itemEmbed,
      rnnHiddenLayers,
      sessionLength,
      includeHistory,
      mlpHiddenLayers,
      historyLength
    ).build()
  }

  /**
   * Load an existing SessionRecommender model (with weights).
   *
   * @param path       The path for the pre-defined model.
   *                   Local file system, HDFS and Amazon S3 are supported.
   *                   HDFS path should be like "hdfs://[host]:[port]/xxx".
   *                   Amazon S3 path should be like "s3a://bucket/xxx".
   * @param weightPath The path for pre-trained weights if any. Default is null.
   * @tparam T Numeric type of parameter(e.g. weight, bias). Only support float/double now.
   */
  def loadModel[T: ClassTag](
      path: String,
      weightPath: String = null)(implicit ev: TensorNumeric[T]): SessionRecommender[T] = {
    ZooModel.loadModel(path, weightPath).asInstanceOf[SessionRecommender[T]]
  }
}

File: pyzoo/zoo/models/recommendation/session_recommender.py
#
# Copyright 2018 Analytics Zoo Authors.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

import sys

from zoo.models.common import KerasZooModel
from zoo.models.recommendation import Recommender
from zoo.pipeline.api.keras.layers import *
from zoo.pipeline.api.keras.models import *
from zoo.common.utils import callZooFunc

if sys.version >= '3':
    long = int
    unicode = str


class SessionRecommender(Recommender):
    """
    The Session Recommender model used for recommendation.

    # Arguments
     item_ount: The number of distinct items. Positive integer.
     item_embed: The output size of embedding layer. Positive integer.
     rnn_hidden_layers: Units of hidden layers for the mlp model. Array of positive integers.
     session_length: The max number of items in the sequence of a session
     include_history: Whether to include purchase history. Boolean. Default is true.
     mlp_hidden_layers: Units of hidden layers for the mlp model. Array of positive integers.
     history_length: The max number of items in the sequence of historical purchase
     """

    def __init__(self, item_count, item_embed, rnn_hidden_layers=[40, 20], session_length=0,
                 include_history=False, mlp_hidden_layers=[40, 20], history_length=0,
                 bigdl_type="float"):
        assert session_length > 0, "session_length should align with input features"
        if include_history:
            assert history_length > 0, "history_length should align with input features"
        self.item_count = int(item_count)
        self.item_embed = int(item_embed)
        self.mlp_hidden_layers = [int(unit) for unit in mlp_hidden_layers]
        self.rnn_hidden_layers = [int(unit) for unit in rnn_hidden_layers]
        self.include_history = include_history
        self.session_length = int(session_length)
        self.history_length = int(history_length)
        self.bigdl_type = bigdl_type
        self.model = self.build_model()
        super(SessionRecommender, self).__init__(None, self.bigdl_type,
                                                 self.item_count,
                                                 self.item_embed,
                                                 self.rnn_hidden_layers,
                                                 self.session_length,
                                                 self.include_history,
                                                 self.mlp_hidden_layers,
                                                 self.history_length,
                                                 self.model)

    def build_model(self):
        input_rnn = Input(shape=(self.session_length,))
        session_table = Embedding(self.item_count + 1, self.item_embed, init="uniform")(input_rnn)

        gru = GRU(self.rnn_hidden_layers[0], return_sequences=True)(session_table)
        for hidden in range(1, len(self.rnn_hidden_layers) - 1):
            gru = GRU(self.rnn_hidden_layers[hidden], return_sequences=True)(gru)
        gru_last = GRU(self.rnn_hidden_layers[-1], return_sequences=False)(gru)
        rnn = Dense(self.item_count)(gru_last)

        if self.include_history:
            input_mlp = Input(shape=(self.history_length,))
            his_table = Embedding(self.item_count + 1, self.item_embed, init="uniform")(input_mlp)
            embedSum = KerasLayerWrapper(Sum(dimension=2))(his_table)
            flatten = Flatten()(embedSum)
            mlp = Dense(self.mlp_hidden_layers[0], activation="relu")(flatten)
            for hidden in range(1, len(self.mlp_hidden_layers)):
                mlp = Dense(self.mlp_hidden_layers[hidden], activation="relu")(mlp)
            mlp_last = Dense(self.item_count)(mlp)
            merged = merge(inputs=[rnn, mlp_last], mode="sum")
            out = Activation(activation="softmax")(merged)
            model = Model(input=[input_rnn, input_mlp], output=out)
        else:
            out = Activation(activation="softmax")(rnn)
            model = Model(input=input_rnn, output=out)
        return model

    def recommend_for_user(self, feature_rdd, max_items):
        raise Exception("recommend_for_user: Unsupported for SessionRecommender")

    def recommend_for_item(self, feature_rdd, max_users):
        raise Exception("recommend_for_item: Unsupported for SessionRecommender")

    def predict_user_item_pair(self, feature_rdd):
        raise Exception("predict_user_item_pair: Unsupported for SessionRecommender")

    def recommend_for_session(self, sessions, max_items, zero_based_label):
        """
        recommend for sessions given rdd of samples or list of samples.

        # Arguments
        sessions: rdd of samples or list of samples.
        max_items:   Number of items to be recommended to each user. Positive integer.
        zero_based_label: True if data starts from 0, False if data starts from 1
        :return rdd of list of list(item, probability),
        """
        if isinstance(sessions, list):
            sc = get_spark_context()
            sessions_rdd = sc.parallelize(sessions)
        elif (isinstance(sessions, RDD)):
            sessions_rdd = sessions
        else:
            raise TypeError("Unsupported training data type: %s" % type(sessions))
        results = callZooFunc(self.bigdl_type, "recommendForSession",
                              self.value,
                              sessions_rdd,
                              max_items,
                              zero_based_label)

        if isinstance(sessions, list):
            return results.collect()
        else:
            return results

    @staticmethod
    def load_model(path, weight_path=None, bigdl_type="float"):
        """
        Load an existing SessionRecommender model (with weights).

        # Arguments
        path: The path for the pre-defined model.
              Local file system, HDFS and Amazon S3 are supported.
              HDFS path should be like 'hdfs://[host]:[port]/xxx'.
              Amazon S3 path should be like 's3a://bucket/xxx'.
        weight_path: The path for pre-trained weights if any. Default is None.
        """
        jmodel = callZooFunc(bigdl_type, "loadSessionRecommender", path, weight_path)
        model = KerasZooModel._do_load(jmodel, bigdl_type)
        model.__class__ = SessionRecommender
        return model
